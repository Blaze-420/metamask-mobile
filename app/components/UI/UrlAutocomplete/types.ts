/**
 * Props for the UrlAutocomplete component
 */

import { Hex } from '@metamask/utils';

// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type UrlAutocompleteComponentProps = {
  /**
   * Callback that is triggered while
   * choosing one of the autocomplete options
   */
  onSelect: (item: AutocompleteSearchResult) => void;
  /**
   * Callback that is triggered while
   * tapping on the background
   */
  onDismiss: () => void;
};

/**
 * Ref for the UrlAutocomplete component
 */
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type UrlAutocompleteRef = {
  /**
   * Search for autocomplete results
   */
  search: (text: string) => void;
  /**
   * Hide the autocomplete results
   */
  hide: () => void;
  /**
   * Show the autocomplete results
   */
  show: () => void;
};

/**
 * The result of a Fuse search
 */
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type FuseSearchResult = {
  type: 'sites' | 'recents' | 'favorites';
  url: string;
  name: string;
};

/**
 * The result of a token search
 */
// eslint-disable-next-line @typescript-eslint/consistent-type-definitions
export type TokenSearchResult = {
  type: 'tokens';
  name: string;
  symbol: string;
  address: string;
  chainId: Hex;
  logoUrl?: string;
};

export type AutocompleteSearchResult = FuseSearchResult | TokenSearchResult;
