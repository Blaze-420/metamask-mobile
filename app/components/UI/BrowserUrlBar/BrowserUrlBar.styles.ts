import { StyleSheet } from 'react-native';

import { Theme } from '../../../util/theme/models';

const styleSheet = (params: { theme: Theme }) =>
  StyleSheet.create({
    main: {
      flexDirection: 'row',
      alignItems: 'center',
    },
    text: {
      color: params.theme.colors.text.alternative,
      flex: 1,
      height: 44,
      padding: 0,
      margin: 0,
      paddingLeft: 8,
    },
  });

export default styleSheet;
