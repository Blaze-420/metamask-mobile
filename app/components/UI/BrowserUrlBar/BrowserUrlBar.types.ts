/**
 * BrowserUrlBar props
 */
export type BrowserUrlBarProps = {
  isSecureConnection: boolean;
  onSubmitEditing: (text: string) => void;
  onCancel: () => void;
  onFocus: () => void;
  onBlur: () => void;
  onChangeText: (text: string) => void;
  connectedAccounts: string[];
  activeUrl: string;
};
