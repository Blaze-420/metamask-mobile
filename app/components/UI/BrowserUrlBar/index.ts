export { default } from './BrowserUrlBar';

export { ConnectionType } from './BrowserUrlBar.types';
