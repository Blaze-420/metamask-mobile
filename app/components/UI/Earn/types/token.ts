import { TokenI as BaseTokenI } from '../../Tokens/types';

export interface EarnTokenI extends BaseTokenI {
  tokenBalanceFormatted: string;
  balanceFiat: string;
}
