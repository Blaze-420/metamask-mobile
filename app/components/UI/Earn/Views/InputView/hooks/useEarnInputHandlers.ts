import { useCallback, useEffect, useState } from 'react';
import { TokenI } from '../../../../Tokens/types';
import type { QuickAmount } from '../../../../Ramp/types';
import { EARN_ACTIONS } from '../../../constants';

interface UseEarnInputHandlersProps {
  token?: TokenI;
  action?: (typeof EARN_ACTIONS)[keyof typeof EARN_ACTIONS];
}

export const useEarnInputHandlers = ({
  token,
  action = EARN_ACTIONS.DEPOSIT,
}: UseEarnInputHandlersProps) => {
  const [selectedToken, setSelectedToken] = useState<TokenI | undefined>(token);
  const [inputAmount, setInputAmount] = useState('0');
  const [inputCurrency, setInputCurrency] = useState(token?.symbol || '');
  const [annualRewardRate] = useState('0');
  const [balanceValue] = useState('0');
  const [gasCostImpactWarning] = useState(false);
  const [currentAction] = useState(action);

  // Mock quick amounts for now - this should be calculated based on token balance
  const quickAmounts: QuickAmount[] = [
    { value: 0.25, label: '25%' },
    { value: 0.5, label: '50%' },
    { value: 0.75, label: '75%' },
    { value: 1, label: 'MAX' },
  ];

  const handleQuickAmountPress = useCallback(
    ({ value }: { value: number }) => {
      // TODO: Implement quick amount selection logic
      setInputAmount((value * parseFloat(balanceValue)).toString());
    },
    [balanceValue],
  );

  const handleCurrencySwitch = useCallback(() => {
    // TODO: Implement currency switching logic
  }, []);

  const handleKeypadPress = useCallback((value: string) => {
    setInputAmount(value);
  }, []);

  const handleReviewPress = useCallback(() => {
    // TODO: Implement review logic based on currentAction
  }, []);

  const handleTokenSelect = useCallback((newToken: TokenI) => {
    setSelectedToken(newToken);
    setInputCurrency(newToken.symbol);
  }, []);

  useEffect(() => {
    if (token && !selectedToken) {
      setSelectedToken(token);
      setInputCurrency(token.symbol);
    }
  }, [token, selectedToken]);

  return {
    selectedToken,
    inputAmount,
    inputCurrency,
    quickAmounts,
    handleQuickAmountPress,
    handleCurrencySwitch,
    handleKeypadPress,
    handleReviewPress,
    handleTokenSelect,
    gasCostImpactWarning,
    annualRewardRate,
    balanceValue,
    currentAction,
  };
};
