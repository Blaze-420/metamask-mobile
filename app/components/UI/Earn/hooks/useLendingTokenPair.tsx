import { TokenI } from '../../Tokens/types';
import {
  LENDING_TOKEN_TO_RECEIPT_TOKEN_MAP,
  RECEIPT_TOKEN_TO_LENDING_TOKEN_MAP,
  SUPPORTED_LENDING_TOKENS,
} from '../utils';
import { useEarnTokenDetails } from './useEarnTokenDetails';
import useEarnTokens from './useEarnTokens';

// Pair = Lending token and its related receipt token (e.g. USDC and AETHUSDC)
// Potential improvement: Consider memoizing pairs to avoid unnecessary recalculating.
// Will wait to see if we use this after earn-controller state is pulled.
const useLendingTokenPair = (token: TokenI) => {
  const { getTokenWithBalanceAndApr } = useEarnTokenDetails();

  const result = { lendingToken: {}, receiptToken: {} };

  const isLendingToken = SUPPORTED_LENDING_TOKENS.has(token.symbol);

  const lendingTokens = useEarnTokens({
    includeLendingTokens: true,
  });

  const receiptTokens = useEarnTokens({ includeReceiptTokens: true });

  if (!token?.chainId) return result;

  if (isLendingToken) {
    const relatedReceiptTokenSymbol =
      LENDING_TOKEN_TO_RECEIPT_TOKEN_MAP[token.chainId][token.symbol];
    const relatedReceiptToken = receiptTokens.find(
      (receiptToken) => receiptToken.symbol === relatedReceiptTokenSymbol,
    );

    if (!relatedReceiptToken) return result;

    result.lendingToken = getTokenWithBalanceAndApr(token);
    result.receiptToken = getTokenWithBalanceAndApr(relatedReceiptToken);

    return result;
  }

  // Find matching token for receipt token
  const relatedLendingTokenSymbol =
    RECEIPT_TOKEN_TO_LENDING_TOKEN_MAP[token.chainId][token.symbol];
  const relatedLendingToken = lendingTokens.find(
    (receiptToken) => receiptToken.symbol === relatedLendingTokenSymbol,
  );

  if (!relatedLendingToken) return result;

  result.lendingToken = getTokenWithBalanceAndApr(relatedLendingToken);
  result.receiptToken = getTokenWithBalanceAndApr(token);

  return result;
};

export default useLendingTokenPair;
