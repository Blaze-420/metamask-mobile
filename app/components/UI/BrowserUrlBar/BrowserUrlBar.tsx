import React, { forwardRef, useMemo, useRef, useState } from 'react';
import {
  NativeSyntheticEvent,
  TextInput,
  TextInputSubmitEditingEventData,
  TouchableOpacity,
  View,
} from 'react-native';
import { useStyles } from '../../../component-library/hooks';
import Icon, {
  IconName,
  IconSize,
} from '../../../component-library/components/Icons/Icon';
import { BrowserUrlBarProps, ConnectionType } from './BrowserUrlBar.types';
import stylesheet from './BrowserUrlBar.styles';
import { BrowserViewSelectorsIDs } from '../../../../e2e/selectors/Browser/BrowserView.selectors';
import { strings } from '../../../../locales/i18n';
import { BrowserURLBarSelectorsIDs } from '../../../../e2e/selectors/Browser/BrowserURLBar.selectors';
import AccountRightButton from '../AccountRightButton';
import Text from '../../../component-library/components/Texts/Text';
import { selectAccountsLength } from '../../../selectors/accountTrackerController';
import { useSelector } from 'react-redux';
import { selectNetworkConfigurations } from '../../../selectors/networkController';
import { useMetrics } from '../../hooks/useMetrics';
import { MetaMetricsEvents } from '../../../core/Analytics';
import { useNavigation } from '@react-navigation/native';
import Routes from '../../../constants/navigation/Routes';
import URLParse from 'url-parse';
// import { getURLProtocol } from '../../../util/general';
// import { PROTOCOLS } from '../../../constants/deeplinks';
// import { isGatewayUrl } from '../../../lib/ens-ipfs/resolver';
// import AppConstants from '../../../core/AppConstants';
// import Url from 'url-parse';
// import { regex } from '../../../../app/util/regex';

const BrowserUrlBar = forwardRef<TextInput, BrowserUrlBarProps>(
  (
    {
      connectionType,
      onSubmitEditing,
      onCancel,
      onFocus,
      onBlur,
      onChangeText,
      connectedAccounts,
      activeUrlRef,
    },
    ref,
  ) => {
    const inputValueRef = useRef<string>('');
    const [isUrlBarFocused, setIsUrlBarFocused] = useState(false);
    const accountsLength = useSelector(selectAccountsLength);
    const networkConfigurations = useSelector(selectNetworkConfigurations);
    const { trackEvent, createEventBuilder } = useMetrics();
    const navigation = useNavigation();
    const selectedAddress = connectedAccounts?.[0];
    const {
      styles,
      theme: { colors, themeAppearance },
    } = useStyles(stylesheet, {});

    // TODO: Decide if this logic is still needed
    // const getDappMainUrl = () => {
    //   if (!url) return;

    //   const urlObj = new Url(url);
    //   const ensUrl = route.params?.currentEnsName ?? '';

    //   if (
    //     isGatewayUrl(urlObj) &&
    //     url.search(`${AppConstants.IPFS_OVERRIDE_PARAM}=false`) === -1 &&
    //     Boolean(ensUrl)
    //   ) {
    //     return ensUrl.toLowerCase().replace(regex.startUrl, '');
    //   }
    //   return urlObj.host.toLowerCase().replace(regex.startUrl, '') || url;
    // };

    // const contentProtocol = getURLProtocol(url);
    // const isHttps = contentProtocol === PROTOCOLS.HTTPS;
    // const mainUrl = getDappMainUrl();

    /**
     * Gets browser url bar icon based on connection type
     */
    const connectionTypeIcon = useMemo(() => {
      // Default to search icon
      let iconName = IconName.Search;
      if (isUrlBarFocused) {
        return iconName;
      }
      switch (connectionType) {
        case ConnectionType.SECURE:
          iconName = IconName.Lock;
          break;
        case ConnectionType.UNSECURE:
          iconName = IconName.LockSlash;
          break;
        case ConnectionType.UNKNOWN:
        default:
          iconName = IconName.Loading;
      }
      return iconName;
    }, [connectionType, isUrlBarFocused]);

    const onBlurInput = () => {
      setIsUrlBarFocused(false);
      onBlur();
      if (!inputValueRef.current) {
        onCancel();
      }
      // Reset the input value
      inputValueRef.current = '';
    };

    const onFocusInput = () => {
      setIsUrlBarFocused(true);
      onFocus();
    };

    const onSubmitEditingInput = ({
      nativeEvent: { text },
    }: NativeSyntheticEvent<TextInputSubmitEditingEventData>) => {
      const trimmedText = text.trim();
      inputValueRef.current = trimmedText;
      onSubmitEditing(trimmedText);
    };

    const handleAccountRightButtonPress = () => {
      const nonTestnetNetworks = Object.keys(networkConfigurations).length + 1;
      const numberOfConnectedAccounts = connectedAccounts.length;

      // TODO: This is currently tracking two events, we should consolidate
      trackEvent(
        createEventBuilder(MetaMetricsEvents.OPEN_DAPP_PERMISSIONS)
          .addProperties({
            number_of_accounts: accountsLength,
            number_of_accounts_connected: numberOfConnectedAccounts,
            number_of_networks: nonTestnetNetworks,
          })
          .build(),
      );
      // Track Event: "Opened Acount Switcher"
      trackEvent(
        createEventBuilder(MetaMetricsEvents.BROWSER_OPEN_ACCOUNT_SWITCH)
          .addProperties({
            number_of_accounts: numberOfConnectedAccounts,
          })
          .build(),
      );

      navigation.navigate(Routes.MODAL.ROOT_MODAL_FLOW, {
        screen: Routes.SHEET.ACCOUNT_PERMISSIONS,
        params: {
          hostInfo: {
            metadata: {
              // TODO: This is not an origin, it's a hostname
              origin:
                activeUrlRef.current &&
                new URLParse(activeUrlRef.current).hostname,
            },
          },
        },
      });
    };

    return (
      <View style={styles.browserUrlBarWrapper}>
        <View style={styles.main} testID={BrowserViewSelectorsIDs.URL_INPUT}>
          <Icon
            color={colors.icon.alternative}
            name={connectionTypeIcon}
            size={IconSize.Sm}
          />
          <TextInput
            testID={BrowserURLBarSelectorsIDs.URL_INPUT}
            keyboardType={'web-search'}
            autoCapitalize={'none'}
            autoCorrect={false}
            ref={ref}
            numberOfLines={1}
            placeholder={strings('autocomplete.placeholder')}
            placeholderTextColor={colors.text.muted}
            returnKeyType={'go'}
            selectTextOnFocus
            keyboardAppearance={themeAppearance}
            style={styles.textInput}
            onChangeText={onChangeText}
            onSubmitEditing={onSubmitEditingInput}
            onBlur={onBlurInput}
            onFocus={onFocusInput}
          />
        </View>
        <View style={styles.rightButton}>
          {isUrlBarFocused ? (
            <TouchableOpacity
              hitSlop={{ top: 12, bottom: 12, left: 12, right: 12 }}
              style={styles.cancelButton}
              testID={BrowserURLBarSelectorsIDs.CANCEL_BUTTON_ON_BROWSER_ID}
              // TODO: Change this to cancel
              onPress={onBlurInput}
            >
              <Text style={styles.cancelButtonText}>
                {strings('browser.cancel')}
              </Text>
            </TouchableOpacity>
          ) : (
            <AccountRightButton
              selectedAddress={selectedAddress}
              isNetworkVisible
              onPress={handleAccountRightButtonPress}
            />
          )}
        </View>
      </View>
    );
  },
);

export default BrowserUrlBar;
