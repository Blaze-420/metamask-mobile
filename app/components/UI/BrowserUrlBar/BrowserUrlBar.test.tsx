import React from 'react';
import renderWithProvider from '../../../util/test/renderWithProvider';
import BrowserUrlBar from './BrowserUrlBar';
import { ConnectionType } from './BrowserUrlBar.types';
import { useMetrics } from '../../../components/hooks/useMetrics';
import { backgroundState } from '../../../util/test/initial-root-state';
import { selectAccountsLength } from '../../../selectors/accountTrackerController';
import { selectNetworkConfigurations } from '../../../selectors/networkController';

const mockNavigate = jest.fn();

jest.mock('@react-navigation/native', () => {
  const actualNav = jest.requireActual('@react-navigation/native');
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockNavigate,
    }),
  };
});

jest.mock('../../../components/hooks/useMetrics', () => ({
  useMetrics: jest.fn(),
}));

const mockTrackEvent = jest.fn();
const mockCreateEventBuilder = jest.fn(() => ({
  addProperties: jest.fn(() => ({
    build: jest.fn(),
  })),
}));

const mockInitialState = {
  engine: {
    backgroundState,
  },
};

const mockUseSelector = jest.fn();
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useSelector: (callback: unknown) => mockUseSelector(callback),
}));

describe('BrowserUrlBar', () => {
  const defaultProps = {
    connectionType: ConnectionType.SECURE,
    onSubmitEditing: jest.fn(),
    onCancel: jest.fn(),
    onFocus: jest.fn(),
    onBlur: jest.fn(),
    onChangeText: jest.fn(),
    connectedAccounts: ['0x123'],
    activeUrl: 'https://example.com',
    setIsUrlBarFocused: jest.fn(),
    isUrlBarFocused: true,
  };

  beforeEach(() => {
    jest.clearAllMocks();
    (useMetrics as jest.Mock).mockReturnValue({
      trackEvent: mockTrackEvent,
      createEventBuilder: mockCreateEventBuilder,
    });

    mockUseSelector.mockImplementation((selector) => {
      if (selector === selectAccountsLength) return 1;
      if (selector === selectNetworkConfigurations) return {};
      return null;
    });
  });

  it('should render correctly', () => {
    const { toJSON } = renderWithProvider(<BrowserUrlBar {...defaultProps} />, {
      state: mockInitialState,
    });
    expect(toJSON()).toMatchSnapshot();
  });
});
