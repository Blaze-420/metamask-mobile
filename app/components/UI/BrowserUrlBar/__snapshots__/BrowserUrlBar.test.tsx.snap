// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrowserUrlBar should render correctly 1`] = `
<View
  style={
    {
      "alignItems": "center",
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "backgroundColor": "#f2f4f6",
        "borderRadius": 999,
        "flex": 1,
        "flexDirection": "row",
        "marginHorizontal": 16,
      }
    }
    testID="url-input"
  >
    <View
      style={
        {
          "flex": 1,
        }
      }
    >
      <TextInput
        autoCapitalize="none"
        autoCorrect={false}
        keyboardAppearance="light"
        keyboardType="web-search"
        numberOfLines={1}
        onBlur={[Function]}
        onChangeText={[Function]}
        onFocus={[Function]}
        onSubmitEditing={[Function]}
        placeholder="Search or Type URL"
        placeholderTextColor="#9fa6ae"
        returnKeyType="go"
        selectTextOnFocus={true}
        style={
          {
            "color": "#141618",
            "flex": 1,
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 14,
            "fontWeight": "400",
            "height": 44,
            "margin": 0,
            "paddingLeft": 16,
            "paddingVertical": 0,
          }
        }
        testID="browser-modal-url-input"
      />
    </View>
    <TouchableOpacity
      accessible={true}
      activeOpacity={1}
      disabled={false}
      hitSlop={
        {
          "bottom": 8,
          "left": 8,
          "right": 8,
          "top": 8,
        }
      }
      onPress={[Function]}
      onPressIn={[Function]}
      onPressOut={[Function]}
      style={
        {
          "alignItems": "center",
          "borderRadius": 8,
          "height": 28,
          "justifyContent": "center",
          "marginLeft": 4,
          "marginRight": 8,
          "opacity": 1,
          "width": 28,
        }
      }
    >
      <SvgMock
        color="#6a737d"
        height={20}
        name="CircleX"
        style={
          {
            "height": 20,
            "width": 20,
          }
        }
        width={20}
      />
    </TouchableOpacity>
  </View>
  <View
    style={
      {
        "height": 50,
        "justifyContent": "center",
      }
    }
  >
    <TouchableOpacity
      hitSlop={
        {
          "bottom": 12,
          "left": 12,
          "right": 12,
          "top": 12,
        }
      }
      onPress={[Function]}
      style={
        {
          "justifyContent": "center",
          "marginRight": 16,
        }
      }
      testID="cancel-url-button"
    >
      <Text
        accessibilityRole="text"
        style={
          {
            "color": "#0376c9",
            "fontFamily": "EuclidCircularB-Regular",
            "fontSize": 14,
            "fontWeight": "400",
            "letterSpacing": 0,
            "lineHeight": 22,
          }
        }
      >
        Cancel
      </Text>
    </TouchableOpacity>
  </View>
</View>
`;
